Forbids use of --insecure in FIPS mode

tcsc opened the pull request:
Addresses #5073

1. Forbids use of `--insecure` in FIPS mode in `teleport`
2. disables the `--insecure` `tsh` command line option

In order to do (2), this change adds the ability to discriminate between FIPS/non-FIPS build in `modules`, via build constraints that supply different versions of `defaultModules.IsBoringBinary()` based on the pesence or absence of the `fips` build tag.

tcsc committed:
Forbids use of --insecure in FIPS mode

Addresses #5073

tcsc commented:
@klizhentas, I just realised I misread ticket #5073 and have patched in `teleport` rather than `tsh`.

Is there a FIPS mode for `tsh` at the moment (I don't _think_ there is, but could be wrong)? Or would fixing this require changing `tsh` build such that `tsh` knows about fips settings/build flags/whatever?

tcsc committed:
Disables --insecure option in tsh FIPS builds

Adds the ability to discriminate FIPS/non-FIPS build in `modules`, via
build constraints that supply different versions of
`defaultModules.IsBoringBinary()` based on the pesence or absence of the
`fips` build tag.

This is then used to enable or disable the `--insecure` `tsh` command
line option.

nklaassen commented on lib/modules/modules.go:
```diff
@@ -116,10 +116,8 @@ func (p *defaultModules) Features() Features {
-func (p *defaultModules) IsBoringBinary() bool {
-	return false
-}
+// (p *defaultModules) IsBoringBinary() implemented in the appropriate
+// is_boring_binary_(fips|non_fips).go file
```
Any reason not to do the same as e/tool/modules/modules.go?

tcsc committed:
Simplify approach

Use the same technique as the enterprise version to detect a
BoringCrypto vbiunary, rather than have p[laying around with custom
build flags.

tcsc committed:
Merge branch 'master' into trent/disable-insecure-in-fips-mode
